### PR TITLE
Rework structure

### PR DESCRIPTION
**Pull request by yuanqing-wang**
## Description
Provide a brief description of the PR's purpose here.

## Todos
Notable points that this PR has either accomplished or will accomplish.
  - [ ] TODO 1

## Questions
- [ ] Question1

## Status
- [ ] Ready to go

### CODE REVIEW

File: mtenn/model.py [hunk 1]
```diff
@@ -10,10 +11,65 @@ class Model():
...
+
+        Returns
+        -------
+        """
+
+        return(self.representation(*args, **kwargs))
+
+    def forward(self, comp, *parts):
+        complex_rep = self.get_representation(*comp)
+        parts_rep = [self.get_representation(*p) for p in parts]
```
Comment (kaminow) [inline mtenn/model.py:35]: one thing that I'm a bit unsure on is the argument unpacking here. I have this in because the SchNet model takes the atomic numbers and position vectors as separate arguments, but I don't know if it makes sense to handle that here in the general Model class or if I should push it to `conversion_utils.schnet.get_representation`

Comment (yuanqing-wang) [reply mtenn/model.py:35]: I think it's safe to assume that atomic numbers and positisions are inputs of any equivariant gnn model

Comment (kaminow) [reply mtenn/model.py:35]: yeah definitely, but the e3nn model I've been using takes the input as a single object and extracts the necessary information in its forward call:
https://github.com/e3nn/e3nn/blob/03b40770fa752cfd524321dbb15525e7ccff773b/e3nn/nn/models/gate_points_2101.py#L295

so in order for the code I have here to be compatible with the e3nn model, each data object would have to be passed in as `(data,)` instead of just passing `data`

Comment (yuanqing-wang) [reply mtenn/model.py:35]: I think the best solution would be to provide wrappers for each model, like

```python

class E3NN(OriginalE3NN):
        def forward(self, x, h):
                data = torch.cat([x, h], dim=-1)
                return super().forward(data)
```

something like that

**Review by yuanqing-wang — COMMENTED** [review]
This looks lovely. Thanks!

File: mtenn/model.py (hunk 1 above)
Comment (kaminow) [reply mtenn/model.py]: @yuanqing-wang implemented this in the two newest commits, let me know what you think!

File: mtenn/conversion_utils/schnet.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+                model.num_interactions, model.num_gaussians,
+                model.cutoff, model.max_num_neighbors,model.readout,
+                model.dipole, model.mean, model.std, atomref)
+            super(SchNet, self).__init__(*model_params)
+            self.load_state_dict(model.state_dict())
+
+    def forward(self, data):
+        return(super(SchNet, self).forward(data[0], data[1]))
+
+def get_representation(model):
```
Comment (jchodera) [inline mtenn/conversion_utils/schnet.py:28]: Would it be possible to use a class structure to localize all functions that operate on a model within a class scope, rather than file scope?

**Review by jchodera — COMMENTED** [review]
To identify how best to refactor this code, it's often helpful to work in a "top-down" manner by sketching out what you'd like the API to look like when building and using the model.

For example, I really liked this old example of the Tensorflow 1.0 keras API, which I have annotated:
image
There, you can see what a _use case_ would look like, where each high-level task (grab a dataset, train a model) are compactly expressed, and the "science bits" we may want to modify (in this case, the model architecture) can easily be changed by composing some high-level abstractions in different ways.

Is there a way we can do that here? We generally want to
* Grab a dataset (e.g. the DES350K dataset, docked Mpro data) 
* Partition it into train, validate, test
* Define features (e.g. one-hot element encoding, ligand/receptor annotation), model architecture (e.g. SchNet with cutoff, or an e3nn with cutoff and L value) and strategy (e.g. single snapshot energy, or difference between complex energy and displaced energy)
* Define how to train it (loss function, iterations, etc)
* Train it
* Evaluate how well it did on the validation set
* Save the model
* Load the model
* Use it for inference on new data

File: mtenn/conversion_utils/schnet.py (hunk 2 above)
Comment (yuanqing-wang) [reply mtenn/conversion_utils/schnet.py:28]: I think that would work better!

**Comment by yuanqing-wang** [comment]
I generally agree with the structure @jchodera has proposed but I would be careful trying to provide abstractions for training / dataset partitioning / evaluating. These components might look drastically different from experient to experiment and I think it would make sense to leave these in the scirpts to avoid over-energineering.

**Comment by jchodera** [comment]
There's no need to refactor everything at once, either! Simplifying/streamlining the components that are most critical right now is all that is needed.

**Comment by kaminow** [comment]
@jchodera @yuanqing-wang I think I've implemented the structure you're talking about, and I've updated the README to show how the new interface is used.

**Comment by yuanqing-wang** [comment]
Looks great. Thanks!

**Comment by yuanqing-wang** [comment]
I think this looks great!

**Review by jchodera — APPROVED** [review]
Looks like a great start! Feel free to merge and forge ahead.
